sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -176,6 +176,8 @@ object ReorderAssociativeOperator extends Rule[LogicalPlan] {
 object OptimizeIn extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     case q: LogicalPlan => q transformExpressionsDown {
+      case expr @ In(v, _) if expr.isListEmpty =>
+        If(IsNull(v), Literal.create(null, BooleanType), FalseLiteral)
```

Comment:
Use `Coalesce`?

Reply:
sorry, but I can't understand your suggestion: `Coalesce` returns the first non-null value. Here we should return `Null` when the value is null, `false` otherwise. I can't think of a function doing this.

Reply:
If `v` is not nullable, we should return false.

Reply:
True. The current conversion does not help the perf. We just need to convert it to `false`, if we know the left side is not nullable.

Reply:
BTW, we should submit a separate PR for this optimizer change. 

We need to backport the fix to 2.2

Reply:
But if we don't change the plan here, then maybe it's worth to keep the initial change in the `buildFilters` to return false there without actually evaluating the filter itself, which is not needed in that case. What do you think?

Reply:
Should I also create a JIRA for the optimizer change then?

Reply:
Based on the SQL standard, the original fix is wrong. More importantly, the fix does not bring any noticeable perf improvement, because `buildFilter` is only used for partition pruning. In the future, we might enhance it for more advanced statistic-based filter inference. For example, foldable expressions can be evaluated earlier and this code change could cause a regression. 

Yes. Please open a new JIRA for optimizer enhancement.